nodes-lib/src/publish.ts

```diff
@@ -1,9 +1,55 @@
 import { type NodeIDs } from "@desci-labs/desci-codex-lib";
-import { getDpidHistory } from "./api.js";
-import { dpidPublish, hasDpid } from "./chain.js";
+import { 
+  IndexedNodeVersion,
+  getDpidHistory,
+  getDraftNode,
+  getLegacyHistory,
+  prePublishDraftNode,
+} from "./api.js";
+import {
+  dpidPublish,
+  hasDpid,
+} from "./chain.js";
 import { codexPublish } from "./codex.js";
 import { Signer } from "ethers";
 import { type DID } from "dids";
+import { NoSuchEntryError } from "./errors.js";
+
+/**
+ * Publish node to Codex, potentially migrating history from dPID token.
+ * Does *not* automatically register a dPID in the alias registry.
+ */
+export const publish = async (
+  uuid: string,
+  didOrSigner: DID | Signer,
+) => {
+  const node = await getDraftNode(uuid);
+  const prepubResult = await prePublishDraftNode(uuid);
+  const manifestDpid = node.manifestData.dpid?.id;
+
+  const hasDpidInManifest = manifestDpid !== undefined;
+  const hasStreamOnRecord = prepubResult.ceramicStream !== null;
+  const shouldDoMigration = hasDpidInManifest && !hasStreamOnRecord;
```

Comment:
should also potentially check if the dpid API returns a match for this CID, relying on manifest for dpid is a good starting point but it could be wrong. it seems we have the history functions in nodes-lib to do the check

Reply:
Call summary: this is difficult partly because CID's aren't unique, but mostly because the CID can change right before this as part of the prepublish step and hence not be in sync with the registry / dpid.org API.

There are some drawbacks of this approach, but we can detect it happening and fix it retroactively if so.